### PR TITLE
[SPARK-28089][SQL] File source v2: support reading output of file streaming Sink

### PR DESCRIPTION
**Pull request by gengliangwang**
## What changes were proposed in this pull request?

File source V1 supports reading output of FileStreamSink as batch. https://github.com/apache/spark/pull/11897
We should support this in file source V2 as well. When reading with paths, we first check if there is metadata log of FileStreamSink. If yes, we use `MetadataLogFileIndex` for listing files; Otherwise, we use `InMemoryFileIndex`.

## How was this patch tested?

Unit test

### CODE REVIEW

**Comment by gengliangwang** [comment]
This PR also resolve several TODO items in https://github.com/apache/spark/pull/24830.

**Comment by gengliangwang** [comment]
retest this please.

**Comment by gengliangwang** [comment]
@cloud-fan @dongjoon-hyun @HyukjinKwon @jose-torres

File: sql/core/src/test/scala/org/apache/spark/sql/streaming/StreamingDeduplicationSuite.scala [hunk 1]
```diff
@@ -197,32 +197,30 @@ class StreamingDeduplicationSuite extends StateStoreMetricsTest {
...
-            q.processAllAvailable()
-            checkDataset(spark.read.parquet(outputPath).as[String], "a")
-
-            inputData.addData("b")
-            q.processAllAvailable()
-            checkDataset(spark.read.parquet(outputPath).as[String], "a", "b")
-          } finally {
-            q.stop()
-          }
+    withTempDir { output =>
```
Comment (gengliangwang) [inline sql/core/src/test/scala/org/apache/spark/sql/streaming/StreamingDeduplicationSuite.scala]: remove `withSQLConf`

File: sql/core/src/test/scala/org/apache/spark/sql/streaming/StreamingQuerySuite.scala [hunk 2]
```diff
@@ -995,59 +995,56 @@ class StreamingQuerySuite extends StreamTest with BeforeAndAfter with Logging wi
 
       // Reading a file sink output in a batch query should detect the legacy _spark_metadata
       // directory and throw an error
-      // TODO: test file source V2 as well.
-      withSQLConf(SQLConf.USE_V1_SOURCE_READER_LIST.key -> "parquet") {
-        val e = intercept[SparkException] {
-          spark.read.load(outputDir.getCanonicalPath).as[Int]
-        }
-        assertMigrationError(e.getMessage, sparkMetadataDir, legacySparkMetadataDir)
+      val e = intercept[SparkException] {
```
Comment (gengliangwang) [inline sql/core/src/test/scala/org/apache/spark/sql/streaming/StreamingQuerySuite.scala]: This is just removing `withSQLConf` and one TODO comment.

File: sql/core/src/test/scala/org/apache/spark/sql/streaming/FileStreamSinkSuite.scala [hunk 3]
```diff
@@ -512,3 +429,174 @@ class FileStreamSinkSuite extends StreamTest {
...
+class FileStreamSinkV1Suite extends FileStreamSinkSuite {
+  import testImplicits._
+
+  override protected def sparkConf: SparkConf =
+    super
+      .sparkConf
+      .set(SQLConf.USE_V1_SOURCE_READER_LIST, "csv,json,orc,text,parquet")
+      .set(SQLConf.USE_V1_SOURCE_WRITER_LIST, "csv,json,orc,text,parquet")
+
+  test("partitioned writing and batch reading") {
```
Comment (cloud-fan) [inline sql/core/src/test/scala/org/apache/spark/sql/streaming/FileStreamSinkSuite.scala]: can you highlight the difference between this test case and the one in `FileStreamSinkV2Suite`?

**Comment by sundarcse1216** [comment]
Mute the thread

File: sql/core/src/test/scala/org/apache/spark/sql/streaming/FileStreamSinkSuite.scala (hunk 3 above)
Comment (gengliangwang) [reply sql/core/src/test/scala/org/apache/spark/sql/streaming/FileStreamSinkSuite.scala]: For V1 suite, it uses Parquet V1, and it matches `HadoopFsRelation` / `FileScanRDD` to check if the plan is as expected. Also, partition pruning is tested.

For V2 suite, it uses Parquet V2, and it matches  `FileTable` / `BatchScanExec`  to check if the plan is as expected. As partition pruning is not supported in V2 yet, we can't test it.

I can abstract the code if the two cases look duplicated to you.

Comment (gengliangwang) [reply sql/core/src/test/scala/org/apache/spark/sql/streaming/FileStreamSinkSuite.scala]: Or, do you mean changing the test case name?

Comment (cloud-fan) [reply sql/core/src/test/scala/org/apache/spark/sql/streaming/FileStreamSinkSuite.scala]: let's abstract the code, this test case is too long to duplicate.

**Comment by gengliangwang** [comment]
retest this please.

**Comment by gengliangwang** [comment]
retest this please.

**Comment by cloud-fan** [comment]
thanks, merging to master!

**Pull request closed without merging**